src/components/legend/style.js

```diff
@@ -25,6 +25,13 @@ module.exports = function style(s, gd) {
         var layers = Lib.ensureSingle(traceGroup, 'g', 'layers');
         layers.style('opacity', d[0].trace.opacity);
 
+        // Marker vertical alignment
+        var valignFactor = 0;
+        if(gd._fullLayout.legend.valign === 'top') valignFactor = 1.0;
+        if(gd._fullLayout.legend.valign === 'bottom') valignFactor = -1.0;
+        var markerOffsetY = valignFactor * (0.5 * (d[0].lineHeight - d[0].height + 3));
```

Comment:
Nice use of [`lineHeight`](https://github.com/plotly/plotly.js/blob/196153e4e76264b8d84ed3fb75db6fc23d4e7082/src/components/legend/draw.js#L516)